source/api_cc/src/DeepTensor.cc

```diff
@@ -91,16 +109,203 @@ run_model (std::vector<VALUETYPE> &	d_tensor_,
 			    &output_tensors));
   
   Tensor output_t = output_tensors[0];
-  assert (output_t.dims() == 1), "dim of output tensor should be 1";
-  int o_size = output_t.dim_size(0);
-
+  // Yixiao: newer model may output rank 2 tensor [nframes x (natoms x noutdim)]
+  // assert (output_t.dims() == 1), "dim of output tensor should be 1";
   auto ot = output_t.flat<VALUETYPE> ();
+  // this is an Eigen Tensor
+  int o_size = ot.size();
 
   std::vector<VALUETYPE> d_tensor (o_size);
   for (unsigned ii = 0; ii < o_size; ++ii){
     d_tensor[ii] = ot(ii);
   }
-  d_tensor_ = d_tensor;
+  // now we map the type-sorted sel-atom tensor back to original order
+  // first we have to get the type-sorted select map
+  std::vector<int> sel_srt = sel_fwd;
+  select_map<int>(sel_srt, sel_fwd, atommap.get_fwd_map(), 1);
+  // remove those -1 that correspond to discarded atoms
+  std::remove(sel_srt.begin(), sel_srt.end(), -1);
+  // now map the tensor back
+  d_tensor_.resize(o_size);
+  select_map<VALUETYPE>(d_tensor_, d_tensor, sel_srt, odim);
+}
+
+void
+DeepTensor::
+run_model (std::vector<VALUETYPE> &		dglobal_tensor_,
+		  std::vector<VALUETYPE> &	dforce_,
+		  std::vector<VALUETYPE> &	dvirial_,
+		  tensorflow::Session *			session, 
+		  const std::vector<std::pair<std::string, tensorflow::Tensor>> & input_tensors,
+		  const AtomMap<VALUETYPE> &		atommap, 
+		  const std::vector<int> &		sel_fwd,
+		  const int				nghost)
+{
+  unsigned nloc = atommap.get_type().size();
+  unsigned nall = nloc + nghost;
+  if (nloc == 0) {
+    // return empty
+    dglobal_tensor_.clear();
+    dforce_.clear();
+    dvirial_.clear();
+    return;
+  }
+
+  std::vector<Tensor> output_tensors;
+  deepmd::check_status (session->Run(input_tensors, 
+			    {name_prefix(name_scope) + "o_global_" + model_type, 
+			     name_prefix(name_scope) + "o_force", 
+			     name_prefix(name_scope) + "o_virial"},
+			    {}, 
+			    &output_tensors));
```

Comment:
Please notice  #812  the two run models should eval the same output tensors.

Reply:
Solved. Now both would call the same run model.